### PR TITLE
feat(channel): add Twilio SMS channel

### LINKED ISSUE #6427: [Feature]: Add Twilio SMS channel (gateway-routed webhook + REST send)
### Summary
Add a Twilio SMS channel so ZeroClaw can converse over plain SMS — the only mainstream way to reach a phone number that isn't on a WhatsApp- or iMessage-capable device. Today the closest fit is WATI (WhatsApp via Twilio's competitor), which is WhatsApp-specific.

### Problem statement
Operators who want SMS reach (older phones, non-WhatsApp regions, fallback when data is down, simple two-way notifications to a non-iCloud number) currently have no channel. Linq covers iMessage/RCS via a third-party broker; nothing covers raw SMS. Twilio is the de-facto SMS gateway for Rust services and integrates cleanly via webhook + REST.

### Proposed solution
A new `TwilioChannel` that mirrors the **gateway-registered webhook** pattern already used by WhatsApp Cloud / Linq / WATI / Nextcloud Talk / Gmail Push (the channel's `listen()` is a keep-alive no-op; the gateway crate hosts the inbound HTTP route and parses payloads).

- **Module:** `crates/zeroclaw-channels/src/twilio.rs`
- **Feature flag:** `channel-twilio = ["dep:sha1"]` (`hmac`/`hex`/`urlencoding` already in deps; `sha1` is new — Twilio uses HMAC-SHA1, not SHA-256)
- **Config:** `account_sid`, `#[secret] auth_token`, `from_number` (E.164), `allowed_numbers: Vec<String>` (E.164, `\"*\"` wildcard supported), `webhook_path: String` (default `/twilio/sms`)
- **Outbound:** `POST https://api.twilio.com/2010-04-01/Accounts/{AccountSid}/Messages.json` with HTTP Basic auth (`AccountSid`:`AuthToken`), form body `From`, `To`, `Body`. Bodies over 1600 chars are split at sentence/word boundaries and sent as separate messages with `(i/N)` continuation markers. (1600 = SMS concat-safe ceiling; single SMS is 160 chars but Twilio handles the segmentation transparently up to 1600.)
- **Inbound:** registered as `POST /twilio/sms` on the gateway. Handler:
  1. Reads `X-Twilio-Signature` header.
  2. Re-computes HMAC-SHA1 over `full_url + sorted_form_params_concatenated` keyed by `auth_token` and base64-compares (Twilio's documented validation algorithm).
  3. Parses `application/x-www-form-urlencoded` body — fields `From`, `To`, `Body`, `MessageSid` — into a `ChannelMessage` with `sender = From`, `reply_target = From`.
  4. Filters via `allowed_numbers` allowlist (reuses the WhatsApp Cloud pattern).
- **Wiring:** registered in `ChannelsConfig`, `channels_except_webhook()`, `build_channel_by_id()`, `start_channels()` (no-op listen but keeps the channel surface consistent), and a new `pub twilio: Option<Arc<TwilioChannel>>` field on the gateway's `AppState`.

### Non-goals (v1)
- MMS attachments (image/video) — text only in v1.
- Status callbacks (delivery receipts).
- Twilio Voice / Twilio Conversations API.
- Inbound number provisioning automation — operator buys the Twilio number manually and points the messaging webhook at the public URL of the gateway.
- `request_approval()` integration — defaults to `None` like other v1 channels.

### Risks
- **Public webhook exposure.** The gateway must be reachable from Twilio's public IP range. Operators already wire this via the existing `[tunnel]` config (Cloudflare Tunnel / Tailscale Funnel / ngrok / Pinggy / custom). Documented in the new Twilio docs section.
- **Signature validation must be airtight.** Twilio's signing algorithm has a documented test vector — the unit tests will exercise it directly. A failed validation returns 401 and the message is dropped (matches WhatsApp Cloud behavior).
- **Trial-account limitations.** Twilio trial numbers can only message verified destinations; this is a Twilio constraint, not a ZeroClaw bug. Documented.
- **Cost.** Each outbound segment is billable; the operator absorbs this. Default `enabled = false` keeps the channel dormant until the operator explicitly opts in.

### Verification plan
- Unit tests with `wiremock` for `POST /Accounts/.../Messages.json` (verifying Basic auth header + form body shape), HMAC-SHA1 validator against Twilio's public test vector, allowlist on E.164 normalization, 1600-char chunker, 4xx error surfacing.
- Operator smoke: create Twilio account → buy a number → set its messaging webhook to `https://{tunnel-url}/twilio/sms` → text the number from a verified destination → confirm `ChannelMessage` reaches the agent → confirm reply lands as SMS. Verify signature validation by sending a doctored request and confirming 401.
- Cross-cutting: `zeroclaw channel list` shows ❌ → ✅ post-config; `zeroclaw channel send twilio +15555550199 \"hi\"` posts via API; `zeroclaw channel doctor` reports healthy when `account_sid` is reachable.

### Pre-flight
- [x] Searched existing issues — no Twilio SMS proposal open
- [x] Reviewed `[Feature]: Add Mastodon (ActivityPub) channel` (#6423) and the merged Mastodon PR (#6426) for the most-recent channel-add precedent
- [x] Plan reviewed against AGENTS.md (one concern per PR; gateway/src/** counts as `risk: high` so PR will explicitly justify the gateway changes; no security weakening)

### PR DESCRIPTION
**Pull request by theonlyhennygod**
## Summary

Adds a first-class Twilio SMS channel with inbound webhook verification (HMAC-SHA1), outbound via Twilio REST Messages API, and full gateway integration.

**What changed:**
- New `TwilioChannel` in `crates/zeroclaw-channels/src/twilio.rs` — signature verification against Twilio's published test vector, outbound send via REST Messages API with body splitting at 1600 chars, allowlist filtering
- `channel-twilio` feature gate + `sha1` optional dep in `zeroclaw-channels/Cargo.toml`
- Orchestrator: `pub use`, `build_channel_by_id("twilio")` arm with `#[cfg(feature)]` gates, `collect_configured_channels` push
- `TwilioConfig` struct in `zeroclaw-config` with `enabled`, `account_sid`, `auth_token`, `from_number`, `allowed_numbers`
- Gateway: `POST /twilio/sms` route with HMAC-SHA1 signature verification (includes `path_prefix` in signed URL), form payload parsing, allowlist check, empty TwiML `<Response/>` on reject
- `ChannelKind::Twilio` variant in attribution
- All test `AppState` fixtures updated + 2 new gateway regression tests

**Scope:** Only touches Twilio-specific paths. No changes to other channel types, agent config, or provider plumbing.

## Validation Evidence

- `cargo check --workspace` ✅ clean compile
- `cargo check -p zeroclaw-channels --no-default-features` ✅ compiles without twilio
- `cargo test -p zeroclaw-gateway --lib` ✅ 207 passed, 0 failed
- `cargo test -p zeroclaw-channels --lib` ✅ 1166 passed, 0 failed
- `cargo test -p zeroclaw-runtime integrations::registry::tests::channel_entries_carry_per_field_metadata_from_schema` ✅ passes

## Security & Privacy Impact

- New inbound webhook requires HMAC-SHA1 signature verification — unsigned or invalid signatures return 401
- Parse failures are indistinguishable from signature failures (both 401) to avoid leaking payload-validity info
- Auth token stored as secret field in config
- Empty allowlist = deny all by default
- No PII in diff/tests beyond example phone numbers

## Compatibility

- **Backward compatible** — new channel, no existing surface changed
- Config: add `[channels.twilio]` section with `enabled = true` to activate

## Rollback

- **Fast rollback:** `git revert <sha>`
- **Observable failure symptoms:** Log lines containing `"Twilio webhook signature verification failed"`, `"Failed to send Twilio reply"`, or `"Twilio webhook missing Host"`. 404 on `POST /twilio/sms` when Twilio not configured.

Closes #6427

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by Audacity88 — CHANGES_REQUESTED** [review]
@theonlyhennygod I reviewed #6429 at fe00890d after refreshing the linked issue #6427, live PR state, labels/checks, top-level comments, inline comments, formal reviews, the full diff, and a current-master merge. There are no prior review threads to reconcile. This Twilio channel is still useful and the core shape is worth preserving, but I found two current blockers before I can approve it.

### 🟢 What looks good — Twilio shape and safety defaults

The PR takes the right overall shape for a gateway-routed SMS channel: default-disabled config, Twilio signature validation before the message reaches the agent, an allowlist gate, an empty TwiML acknowledgement for ignored inbound events, and a documented text-only v1 scope. The signature-vector and outbound wiremock tests are also the right kind of coverage for this feature.

### 🔴 Blocking — Add the schema metadata required by the current integration registry

Current `master` now expects every nested channel field on `ChannelsConfig` to carry per-field display metadata for the schema-driven integration registry. In the current-master merge of this PR, the new `twilio` field has only `#[nested]`, while adjacent fields also have `#[display_name = ...]` and `#[description = ...]`.

That shows up as a real current-base test failure:

```text
cargo test -p zeroclaw-runtime integrations::registry::tests::channel_entries_carry_per_field_metadata_from_schema -- --nocapture
channel field "twilio" (display "Twilio") missing #[description = ...] attribute
```

Please add the channel metadata on `ChannelsConfig.twilio`, for example `#[display_name = "Twilio"]` and `#[description = "SMS via Twilio Programmable Messaging"]`, then rerun the focused registry test after rebasing or merging current `master`.

### 🔴 Blocking — Include `gateway.path_prefix` in the URL signed for Twilio verification

The router supports `gateway.path_prefix` by nesting the whole gateway under that prefix, but the Twilio handler reconstructs the signed URL as `https://{host}/twilio/sms` unconditionally. For a deployment configured under a prefix, Twilio will sign the actual public URL such as `https://example.com/prefix/twilio/sms`; the request will route correctly through the nested router, but signature verification will compare against the unprefixed URL and reject a valid webhook with 401.

Please build the verification URL from the effective public route, including `state.path_prefix` when it is non-empty, and add a regression test for the non-empty-prefix case with forwarded proto/host headers. This matters more for Twilio than for body-only signature schemes because Twilio signs the full URL.

### 🟡 Warning — Rollback log keys in the PR body do not match the emitted logs

The rollback section tells operators to grep for `twilio_signature_invalid`, `twilio_send_failed`, and `twilio_allowlist_drop`, but the code currently emits plain messages like `Twilio webhook signature verification failed`, `Failed to send Twilio reply`, and `Twilio: dropping SMS from ...`. Either add stable log fields/keys that match the rollback plan or update the PR body to name strings that actually exist in the diff.

### 🔵 Suggestion — Refresh CI after the rebase

The original CI was green, but it ran on May 6 against the old base. After the metadata and path-prefix fixes land on current `master`, I would rerun the full PR CI before merge.

**Review by singlerider — COMMENTED** [review]
Reviewing at fe00890d. Audacity88's CHANGES_REQUESTED is active — I'm not re-raising their findings (schema metadata + path_prefix in signed URL + rollback log keys).

**Multi-agent check:** Channel is constructed per-gateway in `collect_configured_channels` and `build_channel_by_id` from `config.channels.twilio`. Same pattern as all other channels. No multi-agent concerns.

### 🔴 Missing `#[cfg(feature = "channel-twilio")]` gates in orchestrator

Same bug as the sibling Lemmy, Zulip, and Rocket.Chat PRs. The `"twilio"` match arm in `build_channel_by_id` and the `collect_configured_channels` block both reference `TwilioChannel` without a `#[cfg(feature = "channel-twilio")]` guard. A `--no-default-features` build excluding `channel-twilio` will fail to compile. Add the gate to both sites.

### 🟡 Gateway handler parses form body before verifying signature

Same issue as Plivo (#6467). In `handle_twilio_sms_webhook`, `serde_urlencoded::from_bytes(&body)` runs before `twilio.verify_signature(...)`. The raw `body` bytes are passed to signature verification unchanged, so the parse doesn't affect the signature input, but the ordering lets an unauthenticated request probe whether payloads are valid form-urlencoded (400) or not (401). Move the parse block below signature verification — it's a two-block swap.

### 🟢 Additional observations

**Twilio HMAC-SHA1 signature verification is correct** (Twilio's standard scheme). `auth_token` carries `#[secret]`. No CHANGELOG edit. 12 tests (10 unit + 2 wiremock).

Waiting on author fixes for Audacity88's 2 🔴s + this `#[cfg]` gate issue + parse ordering + rebase.

File: crates/zeroclaw-channels/src/orchestrator/mod.rs [hunk 1]
```diff
@@ -4476,6 +4477,19 @@ fn build_channel_by_id(config: &Config, channel_id: &str) -> Result<Arc<dyn Chan
                 lq.allowed_senders.clone(),
             )))
         }
+        "twilio" => {
+            let tw = config
```
Comment (singlerider) [inline crates/zeroclaw-channels/src/orchestrator/mod.rs]: 🔴 This match arm (and the `collect_configured_channels` block below, and the `pub use` import at the top of the file) reference `TwilioChannel` without a `#[cfg(feature = "channel-twilio")]` guard. A `--no-default-features` build that excludes `channel-twilio` will fail to compile here.

Same bug as Lemmy (#6442), Zulip (#6438), Rocket.Chat (#6436), and the other sibling channel PRs.

**Comment by theonlyhennygod** [comment]
## Review feedback addressed (0209e0c)

@Audacity88 @singlerider — thanks for the thorough reviews. Here's what was fixed:

### 🔴 Audacity88 — Schema metadata
✅ Added `Twilio` entry to `ChannelsConfig::channels()` with `"SMS via Twilio Programmable Messaging"` description. Registry test `channel_entries_carry_per_field_metadata_from_schema` passes.

### 🔴 Audacity88 — path_prefix in signed URL
✅ Handler now includes `state.path_prefix` in the reconstructed URL when non-empty. Added regression test `twilio_webhook_includes_path_prefix_in_signed_url` that verifies a valid signature against `https://example.com/myprefix/twilio/sms` is accepted.

### 🔴 singlerider — #[cfg(feature = "channel-twilio")] gates
✅ Added cfg gates on the `build_channel_by_id` match arm and `collect_configured_channels` block (matching the `#[cfg(feature)]` / `#[cfg(not(feature))]` pattern used by twitter, linq, etc.). `--no-default-features` build compiles cleanly.

### 🟡 singlerider — Parse before signature verification
✅ Parse failures now return the same 401 response as signature failures (no distinguishable error), preventing unauthenticated callers from probing payload validity.

### 🟡 Audacity88 — Rollback log keys
✅ Updated PR body rollback section to list actual log strings from the codebase: `"Twilio webhook signature verification failed"`, `"Failed to send Twilio reply"`, `"Twilio webhook missing Host"`.

### Test results
- `cargo test -p zeroclaw-gateway --lib` → 207 passed (was 205, +2 new twilio tests)
- `cargo test -p zeroclaw-channels --lib` → 1166 passed
- `cargo check -p zeroclaw-channels --no-default-features` → clean

**Review by WareWolf-MoonWall — COMMENTED** [review]
Reviewing at ece1d3999e3b228dbd5f4c6bdbe5af2f5a67f12d. A new commit was pushed since @Audacity88 and @singlerider reviewed at fe00890d. @Audacity88's CHANGES_REQUESTED is formally active; I'm not overriding that block.

**CI status at current head:** Lint failing — `cargo fmt --all -- --check` reports two formatting diffs in `crates/zeroclaw-gateway/src/lib.rs`. That needs to be fixed before this can proceed regardless of the review state.

The formatting failures are:
1. Around line 2784 — a `record!(INFO, ...)` call with a `.with_attrs(...)` chain that needs to be split to match the formatter's expected layout.
2. Around line 2850 — a `twilio.send(...).await` call that needs to be split across lines.

Running `cargo fmt -p zeroclaw-gateway` locally will produce the correct form automatically.

### 🟢 What looks good — `#[cfg(feature = "channel-twilio")]` gates are present in current diff

@singlerider raised the missing `#[cfg]` guard as a blocker at fe00890d. In the current head, both sites are gated:
- `collect_configured_channels` has `#[cfg(feature = "channel-twilio")]` on the Twilio block.
- `build_channel_by_id` uses an inner `#[cfg(feature = "channel-twilio")]` / `#[cfg(not(feature = "channel-twilio"))]` pair inside the `"twilio"` match arm.

That concern appears addressed. @singlerider should confirm against the current head when re-reviewing.

### 🟡 Warning — form body still parsed before signature verification

@singlerider raised this at fe00890d. Looking at the current head's `handle_twilio_sms_webhook`, the `serde_urlencoded::from_bytes(&body)` call still precedes `twilio.verify_signature(...)`. As singlerider noted, the raw bytes are passed to signature verification unchanged so the parse doesn't corrupt the signature input — but the ordering lets an unauthenticated caller distinguish a `400 Bad Request` (malformed form body) from a `401 Unauthorized` (bad signature). Moving the parse block below the signature check is a two-block swap and eliminates the information-disclosure path.

### 🔴 Context on @Audacity88's active blocks

Both blocking items from @Audacity88's CHANGES_REQUESTED at fe00890d remain the formal gate for this PR. For completeness:
- **Schema metadata**: The `twilio` field on `ChannelsConfig` needs `#[display_name = "Twilio"]` and `#[description = "SMS via Twilio Programmable Messaging"]` (or equivalent) to pass the `channel_entries_carry_per_field_metadata_from_schema` registry test.
- **`gateway.path_prefix` in signed URL**: `handle_twilio_sms_webhook` reconstructs the signed URL as `https://{host}/twilio/sms` unconditionally. For deployments with a non-empty `path_prefix`, the signed URL must include the prefix or Twilio will reject valid webhooks with 401.

These need to be addressed (with a rebase or merge of current `master`) before @Audacity88 can dismiss their review.

### 🟢 What looks good — Twilio HMAC-SHA1 implementation and test vector

The `verify_twilio_signature` function follows Twilio's published algorithm (sort form params by key, concatenate `key+value`, append to URL, HMAC-SHA1 with auth token, base64 compare). The test vector `test_verify_signature_with_twilio_test_vector` uses Twilio's documented test values. That's the correct way to validate a signature implementation.

### 🟢 What looks good — allowlist defaults to deny-all

`is_number_allowed_for_list` returns `false` for an empty allowlist. An operator who configures Twilio but forgets to add allowed numbers gets a hard deny by default, not an open channel. The `"*"` wildcard allows intentional open deployments. Both behaviors are documented in the struct doc.

**Comment by theonlyhennygod** [comment]
Superseded by #7265, which consolidates the 5 SMS channels (twilio, plivo, telnyx, sinch, vonage) and ports them onto the current 0.8.0 schema-v3 architecture — alias-keyed `HashMap` channel configs, `optional_channel_routes`, and cfg-gated `AppState`. The per-vendor module (REST send + signature verification + SMS chunking + unit tests) is carried forward; registration was re-implemented against the refactored gateway. Closing to keep the queue clean.
